File: TOP100/79.java
class Solution {
    public boolean exist(char[][] board, String word) {
        char[] w = word.toCharArray();
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {     
                if (exist(board, i ,j, w, 0)) {
                    return true;
                }   
            }
        }
        
        return false;
    }
    
    private boolean exist(char[][] board, int i, int j, char[] w, int k) {
        if (k == w.length) {
            return true;
        }
        else if (i < 0 || j < 0 || i == board.length || j == board[i].length) {
            return false;
        }
        
        if (board[i][j] != w[k]) {
            return false;
        }
        
        board[i][j] ^= 256;
        
        boolean exist = exist(board, i, j+1, w, k+1)
            || exist(board, i, j-1, w, k+1)
            || exist(board, i+1, j, w, k+1)
            || exist(board, i-1, j, w, k+1);
        
        board[i][j] ^= 256;
        
        return exist;
    }
}

File: 2024_april/79.js
var exist = function (board, word) {

    let visited = [];
    let index = 0;

    for (let i = 0; i < board.length; i++) {
        let row = [];

        for (let j = 0; j < board[0].length; j++) {
            row.push(false);
        }

        visited.push(row);
    }

    for (let i = 0; i < board.length; i++) {
        for (let j = 0; j < board[0].length; j++) {
            let result = backtrack(board, visited, i, j, word, index);

            if (result) {
                return true;
            }
        }
    }

    return false;
};

function backtrack(board, visited, i, j, w, index) {

    if (index >= w.length) {
        return true;
    }

    else if (i < 0 || j < 0 || i == board.length || j == board[i].length) {
        return false;
    }

    if (visited[i][j] || board[i][j] != w.charAt(index)) {
        return false;
    }

    visited[i][j] = true;

    let exists = backtrack(board, visited, i - 1, j, w, index + 1)
        || backtrack(board, visited, i, j - 1, w, index + 1)
        || backtrack(board, visited, i + 1, j, w, index + 1)
        || backtrack(board, visited, i, j + 1, w, index + 1);

    if (exists) {
        return true;
    }

    visited[i][j] = false;

    return false;
}